odoo/orm/models.py

```diff
@@ -2517,7 +2502,17 @@ def _read_group_format_result(self, rows_dict, lazy_groupby):
                         # date/datetime field value to False.
                         row.setdefault('__range', {})[group] = False
 
-                row['__domain'] = expression.AND([row['__domain'], additional_domain])
+                row['__domain'].extend(additional_domain)
+        # normalize resulting '__domain',
+        # it is already a concatenation of normalized domains
+        for row in rows_dict:
+            domain = row['__domain']
+            condition_count = len(domain) - sum(
+                1 if d == '!' else 2
+                for d in domain
+                if isinstance(d, str)
+            )
+            row['__domain'] = ['&'] * (condition_count - 1) + domain
```

Comment:
Instantiation of domain is slower than instantiation of the list, building it like this is much faster.
That function is soon to be deprecated, so that complexity will disappear.

Reply:
> Instantiation of domain is slower than instantiation of the list, building it like this is much faster.
That function is soon to be deprecated, so that complexity will disappear.

Then leave at be. Don't modify it if you don't need to.

Reply:
> Instantiation of domain is slower than instantiation of the list, building it like this is much faster.

How much?  Please give some data about this.  Is the optimization really relevant at the request level?

Reply:
It adds 3us per row (not much). : `env['res.users'].read_group([('id', '>', 5)], [], ['write_date'])` takes a few milliseconds (8ms for a res.users table of 20k rows) (depending on the size of the table) on my machine. Returning hundreds of rows could add a few more milliseconds.

```python
from odoo.fields import Domain

def build_list(domain):
    condition_count = len(domain) - sum(
        1 if d == '!' else 2
        for d in domain
        if isinstance(d, str)
    )
    return ['&'] * (condition_count - 1) + domain

def build_domain(domain):
    return list(Domain(domain))


dom = [('a', '=', 5), ('b', 'in', [1, 2, 3])]
%timeit build_list(dom)  # 270ns
%timeit build_domain(dom)  # 3270ns
```

This function will be deprecated. So either I keep it as-is and it will disappear or I simplify it and it is slower when it returns a lot of rows.

Reply:
The extra few milliseconds in _such a case_ will not make it noticeably slower, IMHO.
Go for the simple version, even if slightly slower.

Reply:
Ok, as it will be reimplemented anyways by @ryv-odoo in web.